util/system/path.go

```diff
@@ -220,5 +221,23 @@ func CheckSystemDriveAndRemoveDriveLetter(path string, inputOS string) (string,
 	//
 	// We must return the second element of the split path, as is, without attempting to convert
 	// it to an absolute path. We have no knowledge of the CWD; that is treated elsewhere.
-	return ToSlash(filepath.Clean(parts[1]), inputOS), nil
+	return ToSlash(cleanPath(parts[1]), inputOS), nil
+}
+
+// An adaptation of filepath.Clean to allow an option to
+// retain the trailing slash, on either of the platforms.
+// Returns path with platform specific separators.
+// See https://github.com/moby/buildkit/issues/5249
+func cleanPath(origPath string) string {
+	cleanedPath := filepath.Clean(origPath)
```

Comment:
When the base OS this code runs on is linux and you get a Windows path, `filepath.Clean()` will not do the right thing. This is why I did a `ToSlash()` before cleaning it. See lines 88 and bellow from:

https://go.dev/play/p/ZnAPJ2__nis

Reply:
There is also a large comment starting at line 30, explaining why we need the `ToSlash()` call before `filepath.Clean()`

Reply:
So we absolutely need to take into account the `inputOS`, as on linux, the entire `pth4` from the example is just one path element. If inputOS is Windows, that is an actual path with multiple elements.

Reply:
Sure, as tested here - https://github.com/moby/buildkit/pull/5317#discussion_r1785785139

Reply:
I guess this has not been resolved?

Reply:
Was resolved, tested it and also added unit tests for the cross-build scenarios:

--snip--

Also fixed for cross-building Windows from Linux, that would fail silently:

Repro dockerfile without `RUN`:
```dockerfile
FROM mcr.microsoft.com/windows/nanoserver:ltsc2022
COPY test1.txt /sample/
#RUN type \sample\test1.txt
```

Build log:
```
$ docker buildx build --platform windows/amd64 `
    --builder buildkitd-dev --no-cache --tag=windows-test . `
    --progress plain `
    --output type=local,dest=./output

#6 [2/2] COPY test1.txt /sample/
#6 DONE 0.1s

#7 exporting to client directory
#7 copying files 31B
#7 copying files 230.88MB 0.8s done
#7 DONE 0.8s
```

Checking results, `sample` is a file instead of a directory:
```sh
# before
$ ls -l output/sample
-rw-r--r-- 1 root root 6 Sep 24 08:57 output/sample

# after
$ ls -l output/sample
total 4
-rw-r--r-- 1 root root 6 Sep 24 08:57 test1.txt
```

NOTE: also covered cases like, where platform-specific `filepath.Clean()` won't strip out the `\\` on Linux:
```dockerfile
COPY test1.txt \\sample\\
```